api/simplehash/schemav2.go

```diff
@@ -0,0 +1,198 @@
+package simplehash
+
+// Public go lang implementation of the simplehash RKVST event encoding scheme
+// This lives with the common api because it's job in life is to ensure stable
+// hashing of event data received by an api consumer
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"hash"
+
+	v2assets "github.com/rkvst/go-rkvstcommon-api-gen/assets/v2/assets"
+	"github.com/rkvst/go-rkvstcommon-api-gen/marshalers/simpleoneof"
+	"github.com/zeebo/bencode"
+)
+
+type HasherV2 struct {
+	hasher    hash.Hash
+	marshaler *simpleoneof.Marshaler
+}
+
+func NewHasherV2() HasherV2 {
+	h := HasherV2{
+		hasher:    sha256.New(),
+		marshaler: NewEventMarshaler(),
+	}
+	return h
+}
+
+type HashOptions struct {
+	accumulateHash         bool
+	publicFromPermissioned bool
+	asConfirmed            bool
+}
+
+type HashOption func(*HashOptions)
+
+func WithAccumulate() HashOption {
+	return func(o *HashOptions) {
+		o.accumulateHash = true
+	}
+}
+
+func WithPublicFromPermissioned() HashOption {
+	return func(o *HashOptions) {
+		o.publicFromPermissioned = true
+	}
+}
+
+func WithAsConfirmed() HashOption {
+	return func(o *HashOptions) {
+		o.asConfirmed = true
+	}
+}
+
+// Reset resets the hasher state
+// This is only useful in combination with WithAccumulate
+func (h *HasherV2) Reset() { h.hasher.Reset() }
+
+// HashEvent hashes a single event according to the public event format
+// available to api consumers.
+//
+// Options:
+//   - WithAccumulate callers wishing to implement batched hashing of multiple
+//     events in series should set this. They should call Reset() at their batch
+//     boundaries.
+//   - WithPubliicFromPermissioned should be set if the event is the
+//     permissioned (owner) counter part of a public attestation.
+//   - WithAsConfirmed should be set if the caller is implementing CONFIRMATION
+//     as part of an evidence subsystem implementation. The expectation is that
+//     the caller has a PENDING record to hand, and is in the process of
+//     creating the CONFIRMED record. It is the CONFIRMED record that needs to
+//     be publicly verifiable.
+func (h *HasherV2) HashEvent(event *v2assets.EventResponse, opts ...HashOption) error {
+	o := HashOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	var err error
+
+	// By default, one hash at at time with a reset.
+	if !o.accumulateHash {
+		h.hasher.Reset()
+	}
+
+	if o.publicFromPermissioned {
+		PublicFromPermissionedEvent(event)
+	}
+
+	// If the caller is responsible for evidence confirmation they will have a
```

Comment:
So for forestrie, we want to hash an event that is pending?

so we pretend the forestrie event is confirmed, and hash it, then when it is confirmed the hash will match?

Reply:
forestrie is an evidence subsystem. It will see events as PENDING. It is forestries job to make them CONFIRMED. We are not pretending. Forestrie is CONFIRM'ing by the act of building the trie. The saas db will only get the CONFIRMED (or maybe it will end up being ANCHORED) state once forestrie is satisfied the event is publically verifiable.

So this arangement is just saying 'what will the hash of the verifiable event record be'